efi/preinstall/checks.go

```diff
@@ -0,0 +1,443 @@
+		mandatoryPcrs = append(mandatoryPcrs, internal_efi.SecureBootPolicyPCR)
+	}
+
+	logResults, err := checkFirmwareLogAndChoosePCRBank(tpm, log, mandatoryPcrs)
+	switch {
+	case tpm2.IsTPMError(err, tpm2.AnyErrorCode, tpm2.AnyCommandCode):
+		return nil, &TPM2DeviceError{err}
+	case err != nil:
+		return nil, &TCGLogError{err}
+	}
+
+	// Record the chosen PCR algorithm.
+	result.PCRAlg = logResults.Alg
+
+	// Record errors for non-mandatory PCRs as warnings.
+	for pcr, err := range logResults.pcrErrs() {
+		switch pcr {
+		case 0:
+			err = &PlatformFirmwarePCRError{err}
+			result.Flags |= NoPlatformFirmwareProfileSupport
+		case 1:
+			err = &PlatformConfigPCRError{err}
+			result.Flags |= NoPlatformConfigProfileSupport
+		case 2:
+			err = &DriversAndAppsPCRError{err}
+			result.Flags |= NoDriversAndAppsProfileSupport
+		case 3:
+			err = &DriversAndAppsConfigPCRError{err}
+			result.Flags |= NoDriversAndAppsConfigProfileSupport
+		case 4:
+			err = &BootManagerCodePCRError{err}
+			result.Flags |= NoBootManagerCodeProfileSupport
+		case 5:
+			err = &BootManagerConfigPCRError{err}
+			result.Flags |= NoBootManagerConfigProfileSupport
+		case 7:
+			err = &SecureBootPolicyPCRError{err}
+			result.Flags |= NoSecureBootPolicyProfileSupport
+		}
+		result.Warnings.addErr(err)
+	}
+
+	if virtMode == detectVirtNone {
+		// Only run platform firmware protection checks if we are not in a VM
+		protectedLocalities, err := checkPlatformFirmwareProtections(runChecksEnv, log)
+		switch {
+		case err != nil:
+			mainErr.addErr(&PlatformFirmwareProtectionError{err})
+		case discreteTPM:
+			switch logResults.StartupLocality {
+			case 0:
+				// TPM2_Startup occurred from locality 0. Mark PCR0 as reconstructible
+				// from anything that runs as part of the static OS (only applicable to
+				// discrete TPMs that can be reset independently of the host CPU, which
+				// isn't really meant to be possible).
+				result.Flags |= StartupLocalityNotProtected
+			case 3:
+				// TPM2_Startup occurred from locality 3. Mark PCR0 as reconstructible
+				// from anything that runs as part of the static OS for the reasons stated
+				// above if access to locality 3 isn't protected.
+				switch {
+				case protectedLocalities&tpm2.LocalityThree == 0 && flags&PermitNoDiscreteTPMResetMitigation > 0:
+					result.Flags |= StartupLocalityNotProtected
+				case protectedLocalities&tpm2.LocalityThree == 0:
+					mainErr.addErr(ErrTPMStartupLocalityNotProtected)
+				}
+			case 4:
+				// There were H-CRTM events.  Mark PCR0 as reconstructible from anything that
+				// runs as part of the static OS for the reasons stated above if access to
+				// locality 4 isn't protected.
+				switch {
+				case protectedLocalities&tpm2.LocalityFour == 0 && flags&PermitNoDiscreteTPMResetMitigation > 0:
+					result.Flags |= StartupLocalityNotProtected
+				case protectedLocalities&tpm2.LocalityFour == 0:
+					mainErr.addErr(ErrTPMStartupLocalityNotProtected)
+				}
```

Comment:
this case is not tested

Reply:
We don't have a way to test these yet without adding some code for mocking whic might be a bit tricky, or until go-tpm2 has the ability to perform HCRTM sequences via the simulator API, which is something I'm working on separately (and is not trivial).